File: model-v2/src/aslv2/boxes.py
"""Numpy box helpers (xyxy). Used for offline detector post-processing and tests."""
import numpy as np


def iou(a: np.ndarray, b: np.ndarray) -> float:
    ix0, iy0 = max(a[0], b[0]), max(a[1], b[1])
    ix1, iy1 = min(a[2], b[2]), min(a[3], b[3])
    iw, ih = max(0.0, ix1 - ix0), max(0.0, iy1 - iy0)
    inter = iw * ih
    if inter == 0.0:
        return 0.0
    area_a = (a[2] - a[0]) * (a[3] - a[1])
    area_b = (b[2] - b[0]) * (b[3] - b[1])
    return float(inter / (area_a + area_b - inter))


def nms(boxes: np.ndarray, scores: np.ndarray, iou_thr: float = 0.5) -> list[int]:
    order = list(np.argsort(-scores))
    keep: list[int] = []
    while order:
        i = order.pop(0)
        keep.append(int(i))
        order = [j for j in order if iou(boxes[i], boxes[j]) <= iou_thr]
    return keep

File: app/src/inference/v2/boxes.ts
// xyxy box helpers — direct port of model-v2/src/aslv2/boxes.py.
export type Box = [number, number, number, number];

export function iou(a: Box, b: Box): number {
  const ix0 = Math.max(a[0], b[0]);
  const iy0 = Math.max(a[1], b[1]);
  const ix1 = Math.min(a[2], b[2]);
  const iy1 = Math.min(a[3], b[3]);
  const iw = Math.max(0, ix1 - ix0);
  const ih = Math.max(0, iy1 - iy0);
  const inter = iw * ih;
  if (inter === 0) return 0;
  const areaA = (a[2] - a[0]) * (a[3] - a[1]);
  const areaB = (b[2] - b[0]) * (b[3] - b[1]);
  return inter / (areaA + areaB - inter);
}

/** Greedy NMS. Returns indices into `boxes`, highest score first. */
export function nms(boxes: Box[], scores: number[], iouThr = 0.5): number[] {
  // argsort descending by score (matches numpy argsort(-scores)).
  let order = scores.map((_, i) => i).sort((i, j) => scores[j] - scores[i]);
  const keep: number[] = [];
  while (order.length) {
    const i = order.shift()!;
    keep.push(i);
    order = order.filter((j) => iou(boxes[i], boxes[j]) <= iouThr);
  }
  return keep;
}
